### PR TITLE
feat(#218): tag/renotify で古い通知を上書き + 後続予定を body に集約

### PR DESCRIPTION
**Pull request by kuu13580**
## Summary

Issue #218 — 通知センターに「その trip で次に来る予定」を常に 1 通だけ表示する **rolling next indicator** を導入。tag/renotify による通知集約 + body 情報密度の再設計 + rolling next 特有の複数バグの修正、加えて Staging 検証で判明した既存の UX 不具合 (foreground icon/badge, ボタン loading) を同時に修正。

## 設計思想

**Rolling next indicator**: 通知センターに「次に来る予定」を常に 1 通だけ表示。古い / 時刻変更された通知は自動置換する。未タップ過去通知が消える trade-off は受容 (docs §5b)。

**tag スコープ**:
| 経路 | tag |
|---|---|
| 5 分前通知 (tick) | `trip-{tripId}` |
| テスト送信 | `test-{tripId}` (本運用と分離、テスト直後に本運用で消えないため) |

`tripId` は数値 PK。frontend foreground toast も payload に `tripId` を含めて同一 tag を組む。

### 1. tag / renotify 導入 (`cbceda6`)

- `WebpushNotification` に `tag` / `renotify` 引数追加
- tick: `tag=trip-{tripId}`, `renotify=true` を渡す
- テスト送信: `tag=test-{tripId}`, `renotify=true`
- payload に `tripId` を追加し frontend の tag と統一

### 2. Body / Title の再設計 (`cbceda6`)

- **Title**: `next HH:MM` → `next {block_title}` (5 分前通知は時刻暗黙、block 名の情報量を優先)
- **Body 1 行目**: `▶ HH:MM {block_title}` (現在アラート予定)
- **場所行**: `　📍 {location}` (全角スペース U+3000 でインデント、詳細を視覚的に階層表現)
- **後続予定**: 同一 page 内で next block の絶対時刻以降 (同時刻を含む) を最大 2 件、`▶ HH:MM ブロック名` で追加
- **trip 名を body から削除**: 通知が 1 通に集約されるため識別不要、body 4 行で iOS 4 行制限に fit

### 3. 通知ボタン click 中の loading 表示修正 (`66e816c`)

`useTripSubscription.isLoading` は SWR GET の状態しか反映しないため、click 起因の permission リクエスト / SW 登録 / FCM token 取得 / subscribe API の待ち時間 (体感 3-10 秒) がカバーできていなかった。ローカル `isPending` state を追加し spinner + disabled で連打防止。

### 4. Foreground toast の icon/badge 欠如を修正 (`6b4bfc0`)

Firebase Web SDK の onMessage 経由 (foreground) では payload.notification から自前で `showNotification` を呼ぶが、これまで icon/badge を渡していなかったためブラウザ既定アイコンで表示されていた (前景/背景で見た目が食い違う既存バグ)。`subscribeForegroundMessages` の payload に `icon` を追加、`badge` は Firebase Web SDK 0.13.0 の `NotificationPayload` に無いので同 origin パスを frontend で hardcode。

### 5. 近接時刻の block で先行通知が start 前に上書きされる問題を修正 (`7d003dd`)

シナリオ: A@01:00, B@01:02 で、00:57 tick の B 通知が A を tag で上書きし、**A が start する 01:00 より前に A の通知が消えてしまう** バグ。

対策: `list_notification_candidates` で「同 page 内に candidate より早い upcoming block (`now < 絶対時刻 < candidate 絶対時刻`) が存在するか」を判定し、存在すれば candidate の送信を後 tick に延期する heuristic を追加。A が start (絶対時刻 <= now) すれば次 tick で earliest 判定から抜け、B が新たな next として送信される。A の通知 body には元々 upcoming で B が入っているため、A 表示中に B の存在も認識できる (rolling next 思想と整合)。

### 6. 同時刻 block の情報損失を修正 (`49c8ace`)

end_time null と duration あり block 等で同一 `start_time` は実運用で発生する (frontend UI でも同時刻を semantic にグループ表示)。同時刻 A/B は同一 tag で片方だけが表示され、負けた方の情報が完全に消失していた。

`_collect_upcoming_blocks` の境界を `absolute <= after_utc` (同時刻 drop) から `absolute < after_utc` (同時刻含める) に変更。A の body に B が upcoming で載り、B の body に A が upcoming で載るので、勝者側の body に他方が現れて情報損失を防ぐ。main の割り当ては非決定的だが情報損失ゼロ。

### Backend
- `server/app/firebase.py`: `send_fcm` / `_build_message` に tag / renotify 引数追加
- `server/app/cruds/notification.py`:
  - `NotificationCandidate` に `trip_id` / `upcoming_blocks` 追加
  - `_collect_upcoming_blocks` (同時刻含む)
  - `_has_earlier_upcoming_in_same_page` (延期 heuristic)
- `server/app/notification_content.py`: `format_title` / `format_body` を再設計、後続予定と 📍 インデント
- `server/app/routers/notification_internal.py`: tick 送信で tag / renotify / tripId を渡す
- `server/app/routers/notification.py`: テスト送信で tag / renotify を渡す

### Frontend
- `frontend/src/lib/messaging.ts`: `subscribeForegroundMessages` の payload に `tripId` / `icon` を追加
- `frontend/src/hooks/useForegroundNotificationToast.ts`: `tag=trip-{tripId}`, `renotify: true`, `icon` / `badge` (hardcode) を `showNotification` に渡す
- `frontend/src/components/NotificationToggleButton.tsx`: ローカル `isPending` state で click 中の spinner + 連打防止

### Docs / Tests
- `docs/notifications.md`: §5 通知フォーマット刷新、§5b tag/renotify 設計判断節を新設、既知制約 (iOS renotify best-effort / 順序逆転 / body 長制限 / 同時刻 main 非決定的) を明記
- `server/tests/test_notification_content.py`, `server/tests/cruds/test_notification_upcoming.py` を新規追加
- `frontend/src/components/NotificationToggleButton.test.tsx` に 3 テスト追加 (in-flight spinner / 連打防止 / early return では spinner 出さない)

## 既知制約 (docs/notifications.md §5b)

- iOS Safari (PWA) の renotify は best-effort (tag 置換は効くが再アラートは効かない可能性)
- 同 tick 複数発火時の FCM 順序逆転で古い通知が最終表示される可能性 (実運用で稀)
- プラットフォーム別 body 長制限で末尾切れは許容 (先頭に重要情報を配置)
- 未タップの古い block 通知は消える (rolling next の思想上の trade-off)
- 同時刻 block の「main」割り当ては非決定的 (FCM 配送順序保証なし)。ただし情報は upcoming 経由で両方見える
- 延期 heuristic は same-page のみ判定。cross-page/cross-day は edge case として無対応
- kind 将来拡張 (`at_start` 等) は同一 trip tag で置換される想定 (拡張時に再検討)

## Test plan

- [ ] @coderabbitai review
- [ ] CI (frontend type-check / lint / test / backend pytest) 通過
- [ ] Staging: 5 分前通知で trip 内の旧通知が置換される
- [ ] Staging: iOS PWA で tag 置換が効く (renotify は best-effort)
- [ ] Staging: テスト送信は本運用通知で消えない (別 tag)
- [ ] Staging: 通知ボタン click 中に spinner + disabled が反映される、連打で二重発火しない
- [ ] Staging: foreground toast で icon (オレンジ) と badge (紙飛行機) が正しく表示される
- [ ] Staging: A@10:00 / B@10:02 で B の通知が A の start より前に来ない
- [ ] Staging: 同時刻 A/B で片方消えても body に両方の block 名が残る

## Summary by CodeRabbit

* **新機能**
  * 通知タイトルと本文を改善し、予定時刻、場所、行き先、後続予定を分かりやすく表示。
  * 通知を旅行単位で管理し、関連通知の更新・再通知に対応。
  * フォアグラウンド通知にアイコン、バッジ、旅行情報を追加。

* **改善**
  * 通知処理中はボタンを無効化し、スピナーを表示。二重操作を防止。
  * より早い予定がある場合は通知を延期し、重複を削減。

### CODE REVIEW

File: docs/notifications.md [hunk 1]
```diff
@@ -161,34 +161,90 @@ tick が 60 秒を超えると次 tick と重なる。閾値と対処:
...
+[旅程末尾 — 後続 0 件]
+  Title:  next 就寝
+  Body:   ▶ 21:00 就寝
+          　📍 旅館
 ```
 
-- **Title**: `next HH:MM` 固定 (popup 4 字で "next" が見える、通知センター 10 字で完全表示)
-- **Body**: 改行区切りで **block 名 → 場所行 → trip 名**。省略表示でも block 名が最上部に残る
+- **Title**: `next {block_title}` (Issue #218)。5 分前通知は「これから直近で始まる予定」の意味論なので、時刻より block 名の情報量を優先。通知プレビュー / lock screen で「何の予定」が即座に判る
+  - Title 切れ域: Safari macOS ~34 chars → block 名 ~29 chars 使える。日本語の一般的な block 名 (< 15 chars) はほぼ全環境で完全表示
```
[the remark this answers is not in the document]
Comment (kuu13580) [reply docs/notifications.md:197]: @coderabbitai Title の `next` プレフィックスは意図的な英語表記で、これは仕様として確定済みです (docs §5)。

理由:
- Popup 4 chars 表示で `next` の 4 字が読める (日本語だと 2 字しか表示できず、「予定」等では情報量が減る)
- 通知センター短縮 10 chars 相当でも `next HH:MM` 相当が完全表示
- 日本語 UI 内の英字プレフィックスによる視認性・区別性の向上を設計判断として採用

user とすでに合意済みの設計判断のため、この指摘は却下します。
